### ISSUE TITLE
relationship x is required reports on wrong relationship

### ISSUE THREAD

**Issue by michaelst** [ash-project/ash#433]
**Describe the bug**
I was creating a resource with two different relationships, the first one was valid, the second one was not. However, the error that was returned was that the first relationship was required even though it was valid. As soon as I fixed the invalid second one the error went away.

This was the error `* relationship database is required`, which maybe could also be improved, as the id was provided, it was just invalid. Maybe something like `* relationship database is invalid`

**To Reproduce**
```
changeset
|> Ash.Changeset.manage_relationship(:database, credential.database_id,
  type: :append_and_remove
)
|> Ash.Changeset.manage_relationship(:agent, socket.join_ref, type: :append_and_remove)
|> Api.create!()
```

**Expected behavior**
An error saying the correct relationship was invalid: ``* relationship agent is invalid``

**Runtime**
 - Elixir version: 1.14.1
 - Erlang version: 25.1.2
 - OS: mac
 - Ash version: 2.0.0
 - any related extension versions

**Comment by zachdaniel** [comment]
Are `database` and `agent` both belongs to relationships? This one is definitely confusing, I don't see any way this can happen 😭

**Comment by michaelst** [comment]
Yes they are both belongs_to

**Comment by zachdaniel** [comment]
Would it be possible to reproduce this in a test or a script? I hunted for a while but couldn't see how something like this could be happening.

**Comment by michaelst** [comment]
strangely I can't recreate it in the ash_postgres repo, I am on the latest version. I did see the correct error lower down in the stacktrace as I am trying to reproduce. But I still don't understand where the first error is coming from

```
* relationship database is required
  (elixir 1.14.1) lib/process.ex:773: Process.info/2
  (ash 2.3.0) lib/ash/error/changes/required.ex:5: Ash.Error.Changes.Required.exception/1
  (ash 2.3.0) lib/ash/actions/managed_relationships.ex:356: anonymous fn/2 in Ash.Actions.ManagedRelationships.validate_required_belongs_to/2
  (elixir 1.14.1) lib/enum.ex:2468: Enum."-reduce/3-lists^foldl/2-0-"/3
  (ash 2.3.0) lib/ash/actions/managed_relationships.ex:276: Ash.Actions.ManagedRelationships.setup_managed_belongs_to_relationships/3
  (ash 2.3.0) lib/ash/actions/create.ex:319: anonymous fn/8 in Ash.Actions.Create.as_requests/5
  (ash 2.3.0) lib/ash/changeset/changeset.ex:1460: Ash.Changeset.run_around_actions/2
  (ash 2.3.0) lib/ash/actions/create.ex:318: anonymous fn/8 in Ash.Actions.Create.as_requests/5
  (ash 2.3.0) lib/ash/engine/request.ex:1013: Ash.Engine.Request.do_try_resolve_local/4
  (ash 2.3.0) lib/ash/engine/request.ex:281: Ash.Engine.Request.do_next/1
  (ash 2.3.0) lib/ash/engine/request.ex:210: Ash.Engine.Request.next/1
  (ash 2.3.0) lib/ash/engine/engine.ex:623: Ash.Engine.advance_request/2
  (ash 2.3.0) lib/ash/engine/engine.ex:529: Ash.Engine.fully_advance_request/2
  (ash 2.3.0) lib/ash/engine/engine.ex:470: Ash.Engine.do_run_iteration/2
  (elixir 1.14.1) lib/enum.ex:2468: Enum."-reduce/3-lists^foldl/2-0-"/3
  (ash 2.3.0) lib/ash/engine/engine.ex:224: Ash.Engine.run_to_completion/1
  (ash 2.3.0) lib/ash/engine/engine.ex:169: Ash.Engine.do_run/2
  (ash 2.3.0) lib/ash/engine/engine.ex:89: anonymous fn/2 in Ash.Engine.run/2
  (ecto_sql 3.9.0) lib/ecto/adapters/sql.ex:1190: anonymous fn/3 in Ecto.Adapters.SQL.checkout_or_transaction/4
  (db_connection 2.4.2) lib/db_connection.ex:1562: DBConnection.run_transaction/4
* record with id: "35139a09-87b6-406d-b1d6-939e79f5092a" not found
    at agent
  (elixir 1.14.1) lib/process.ex:773: Process.info/2
  (ash 2.3.0) lib/ash/error/query/not_found.ex:5: Ash.Error.Query.NotFound.exception/1
  (ash 2.3.0) lib/ash/actions/managed_relationships.ex:391: Ash.Actions.ManagedRelationships.create_belongs_to_record/7
  (elixir 1.14.1) lib/enum.ex:4751: Enumerable.List.reduce/3
  (elixir 1.14.1) lib/enum.ex:2514: Enum.reduce_while/3
  (ash 2.3.0) lib/ash/actions/managed_relationships.ex:77: Ash.Actions.ManagedRelationships.setup_managed_belongs_to_relationships/3
  (ash 2.3.0) lib/ash/changeset/changeset.ex:1427: anonymous fn/2 in Ash.Changeset.run_around_actions/2
  (elixir 1.14.1) lib/enum.ex:4751: Enumerable.List.reduce/3
  (elixir 1.14.1) lib/enum.ex:2514: Enum.reduce_while/3
  (ash 2.3.0) lib/ash/changeset/changeset.ex:1404: Ash.Changeset.run_around_actions/2
  (ash 2.3.0) lib/ash/actions/create.ex:318: anonymous fn/8 in Ash.Actions.Create.as_requests/5
  (ash 2.3.0) lib/ash/engine/request.ex:1013: Ash.Engine.Request.do_try_resolve_local/4
  (ash 2.3.0) lib/ash/engine/request.ex:281: Ash.Engine.Request.do_next/1
  (ash 2.3.0) lib/ash/engine/request.ex:210: Ash.Engine.Request.next/1
  (ash 2.3.0) lib/ash/engine/engine.ex:623: Ash.Engine.advance_request/2
  (ash 2.3.0) lib/ash/engine/engine.ex:529: Ash.Engine.fully_advance_request/2
  (ash 2.3.0) lib/ash/engine/engine.ex:470: Ash.Engine.do_run_iteration/2
  (elixir 1.14.1) lib/enum.ex:2468: Enum."-reduce/3-lists^foldl/2-0-"/3
  (ash 2.3.0) lib/ash/engine/engine.ex:224: Ash.Engine.run_to_completion/1
  (ash 2.3.0) lib/ash/engine/engine.ex:169: Ash.Engine.do_run/2
    (elixir 1.14.1) lib/process.ex:773: Process.info/2
    (ash 2.3.0) lib/ash/error/invalid.ex:5: Ash.Error.Invalid.exception/1
    (ash 2.3.0) lib/ash/error/error.ex:239: Ash.Error.choose_error/2
    (ash 2.3.0) lib/ash/actions/create.ex:112: Ash.Actions.Create.do_run/4
    (ash 2.3.0) lib/ash/actions/create.ex:38: Ash.Actions.Create.run/4
    (ash 2.3.0) lib/ash/api/api.ex:944: Ash.Api.create!/3
    (elixir 1.14.1) lib/enum.ex:1658: Enum."-map/2-lists^map/1-0-"/2
    (query_desk 0.1.0) lib/query_desk_web/channels/agent_channel.ex:18: QueryDeskWeb.AgentChannel.handle_info/2
    (phoenix 1.6.15) lib/phoenix/channel/server.ex:343: Phoenix.Channel.Server.handle_info/2
    (stdlib 4.1.1) gen_server.erl:1123: :gen_server.try_dispatch/4
    (stdlib 4.1.1) gen_server.erl:1200: :gen_server.handle_msg/6
    (stdlib 4.1.1) proc_lib.erl:240: :proc_lib.init_p_do_apply/3
Last message: {:after_join, %Websocket.V1.Payload{event: {:phx_join, %{credentials: [%{database_id: "db_01GGKAMFMVMZH843FBWNJ8HY4Z", description: "full access", reviews_required: 0, username: "postgres"}], token: "SFMyNTY.g2gDbQAAAB5hZ3RfMDFHR0tBTUZOMDlNNzRXSDM3RkgzUlNKUEduBgCgPqomhAFiAAFRgA.iUuSN2GeM7xPgyY8zulZv4fUlm2todazbOiVt_fY7W8"}}, __unknown_fields__: []}}
```

**Comment by zachdaniel** [comment]
🤔 I wonder if its something like an error produced managing one relationship causes the next manage relationship not to happen at all, causing the `database` relationship not to be set...

**Comment by michaelst** [comment]
Oh I got it! If you have `allow_nil?` on the belongs_to relationship it throws that error

**Comment by zachdaniel** [comment]
Yeah, thats interesting....I think what we might need to do is if an error happens we should set some context on the changeset like "don't show required errors for these relationships, because we didn't end up trying to manage them"

**Comment by zachdaniel** [comment]
Maybe the simpler thing is  just not to add the final call to `require_fields` if an error is produced when managing relationships.

**Comment by zachdaniel** [comment]
Hm...I think I'm on the wrong path. It may be as simple as only doing this check: https://github.com/ash-project/ash/blob/main/lib/ash/actions/create.ex#L366 if the changeset is vaid

**Comment by zachdaniel** [comment]
@michaelst would you want to test that out/PR it?

**Comment by michaelst** [comment]
https://github.com/ash-project/ash/pull/437
https://github.com/ash-project/ash_postgres/pull/117

**Comment by zachdaniel** [comment]
Resolved in #437
